src/App.js

```diff
@@ -2,41 +2,26 @@ import BookList from './components/BookList'
 import Navigation from './components/Navigation'
 import AddBook from './components/AddBook'
 import Start from './components/Start'
+import usePathname from './hooks/usePathname'
 import styled from 'styled-components/macro'
 import { Route, Switch, useLocation, Redirect } from 'react-router-dom'
 import { useState, useEffect } from 'react'
+import getLocalStorage from './lib/loadFromLocal'
+import setLocalStorage from './lib/saveToLocal'
 
 function App({ data }) {
-  const [books, setBooks] = useState(data)
+  const [username, setUsername] = useState(getLocalStorage('user') || '')
+  const [books, setBooks] = useState(
+    getLocalStorage(`books${username}`) || data
+  )
   const [filteredBooks, setFilteredBooks] = useState(books)
-  const [readingStatus, setReadingStatus] = useState('')
   const { pathname } = useLocation()
-  const [username, setUsername] = useState('')
-
-  function handleBookList(status, books) {
-    setReadingStatus(status)
-    if (status === 'finishedBooks') {
-      const readBooks = books.filter(book => book.finished === true)
-      setFilteredBooks(readBooks)
-    }
-    if (status === 'currentlyReading') {
-      const currentlyReadBooks = books.filter(book => book.finished === false)
-      setFilteredBooks(currentlyReadBooks)
-    }
-    if (status === '') {
-      setFilteredBooks(books)
-    }
-  }
 
   useEffect(() => {
-    setReadingStatus(pathname)
-    if (pathname === '/currently-reading') {
-      handleBookList('currentlyReading', books)
-    }
-    if (pathname === '/library') {
-      handleBookList('finishedBooks', books)
-    }
-  }, [pathname, books])
+    setLocalStorage('user', username)
+  }, [username])
+
+  usePathname(pathname, books, setFilteredBooks)
```

Comment:
this customHook could manage the filteredBooks state as well, and return it like so: 
`const {filteredBooks} = usePathname()`

That said, after digging into it deeper i guess you could use pathname without useEffect, directly in the routes.